onnxruntime/core/providers/webgpu/math/gemm_vec4.cc

```diff
@@ -0,0 +1,323 @@
+  //                           << "  }\n";
+
+  shader.MainFunctionBody() << " mm_write(m, n, values);\n";
+
+  return Status::OK();
+}
+
+bool CanApplyGemmVec4(const Tensor* a,
+                      const Tensor* b,
+                      const Tensor* c) {
+  const auto& a_shape = a->Shape();
+  const auto& b_shape = b->Shape();
+
+  uint32_t A_rows = onnxruntime::narrow<uint32_t>(a_shape[0]);
+  uint32_t A_cols = onnxruntime::narrow<uint32_t>(a_shape[1]);
+  uint32_t B_rows = onnxruntime::narrow<uint32_t>(b_shape[0]);
+  uint32_t B_cols = onnxruntime::narrow<uint32_t>(b_shape[1]);
+
+  if (A_rows % 4 != 0 || A_cols % 4 != 0 || B_rows % 4 != 0 || B_cols % 4 != 0) {
+    return false;
+  }
+
+  // C is optional and is might scalar or 1D.
+  // We do vec4 for C so we need to check if C is 2D and its second dimension is divisible by 4.
+  if (c == nullptr) {
+    return true;
+  }
+
+  const auto& c_shape = c->Shape();
+  if (c_shape.NumDimensions() == 2) {
+    uint32_t C_cols = onnxruntime::narrow<uint32_t>(c_shape[1]);
+    return C_cols % 4 == 0;
+  }
+
+  uint32_t C_rows = onnxruntime::narrow<uint32_t>(c_shape[0]);
+  return C_rows % 4 == 0;
+}
+
+Status ApplyGemmVec4(const Tensor* a,
+                     const Tensor* b,
+                     const Tensor* c,
+                     bool transA_,
+                     bool transB_,
+                     float alpha,
+                     float beta,
+                     ComputeContext& context,
+                     Tensor* y) {
+  const auto& a_shape = a->Shape();
+  const auto& b_shape = b->Shape();
+
+  uint32_t M = onnxruntime::narrow<uint32_t>(transA_ ? a_shape[1] : a_shape[0]);
+  uint32_t K = onnxruntime::narrow<uint32_t>(transA_ ? a_shape[0] : a_shape[1]);
+  uint32_t N = onnxruntime::narrow<uint32_t>(transB_ ? b_shape[0] : b_shape[1]);
+
+  // WebGPU doesn't support binding a zero-sized buffer, so we need to check if A or B is empty.
+  bool need_handle_matmul = a_shape.Size() > 0 && b_shape.Size() > 0;
+  bool need_handle_bias = c && beta;
+
+  GemmVec4Program program{transA_, transB_, alpha, need_handle_bias, need_handle_matmul};
+
+  const int components = 4;
+
+  if (need_handle_matmul) {
+    program.AddInputs({{a, ProgramTensorMetadataDependency::TypeAndRank, components},
+                       {b, ProgramTensorMetadataDependency::TypeAndRank, components}});
+  }
+
+  if (need_handle_bias) {
+    program.AddInput({c, ProgramTensorMetadataDependency::TypeAndRank, components});
+  }
+
+  const uint32_t TILE_SIZE = 32;
+  const uint32_t num_tile_n = (N + TILE_SIZE - 1) / TILE_SIZE;
+  const uint32_t num_tile_m = (M + TILE_SIZE - 1) / TILE_SIZE;
+
+  program.CacheHint(alpha, transA_, transB_)
+      .AddOutputs({{y, ProgramTensorMetadataDependency::TypeAndRank, components}})
+      .SetDispatchGroupSize(num_tile_n * num_tile_m)
+      .SetWorkgroupSize(256, 1, 1)
+      .AddUniformVariables({
+          {static_cast<uint32_t>(num_tile_n)},  // num_tile_n
```

Comment:
`static_cast<uint32_t>` is not needed since `num_tile_n` is type `uint32_t`. Similar for other places.